### PR TITLE
Openshift adm groups sync/prune

### PR DESCRIPTION
**Pull request by abikouo**
Similar to ``oc adm groups sync`` and ``oc adm groups prune`` commands, this new module ``openshift_adm_groups_sync`` add the ability to sync/prune openshift groups with LDAP server based on a sync configuration.

### CODE REVIEW

File: plugins/module_utils/openshift_groups.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
+#!/usr/bin/env python
+
```
Comment (Akasurde) [inline plugins/module_utils/openshift_groups.py:2]: Add License for this file.

**Review by jillr — COMMENTED** [review]
The module and unit tests look good.  I want to re-review plugins/module_utils/openshift_ldap.py tomorrow, it was getting late when I was reviewing it and I want to be sure I had a chance to fully grasp the whole thing.  I will review the integration tests tomorrow.  Good work here @abikouo, thanks!

File: plugins/module_utils/openshift_groups.py [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+        if len(result["resources"]) == 0:
+            return None
+        if len(result["resources"]) == 1 and not return_list:
+            return result["resources"][0]
+        else:
+            return result["resources"]
+
+    def list_groups(self):
+        whitelist = self.module.params.get("whitelist")
+        blacklist = self.module.params.get("blacklist")
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/module_utils/openshift_groups.py:124]: ok will do it, thanks

File: ci/Dockerfile [hunk 3]
```diff
@@ -14,6 +14,8 @@ RUN yum install -y \
       python3-devel \
       python3-pip \
       python3-setuptools \
+      gcc \
```
[the remark this answers is not in the document]
Comment (abikouo) [reply ci/Dockerfile:17]: I dont really have the full picture on how prow retrieves the images used to create the job from which integration testing will be run, but i think we can save directly our image into registry.access.redhat.com, I dont know if we need specific credentials but I will try to deeper investigate

File: plugins/module_utils/openshift_groups.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+            )
+
+        ldapurl_annotation = group["metadata"].get("annotations", {}).get(LDAP_OPENSHIFT_URL_ANNOTATION)
+        if not ldapurl_annotation or ldapurl_annotation != self.module.netlocation:
+            return None, "Group %s: %s annotation did not match sync host: wanted %s, got %s" % (
+                group_name, LDAP_OPENSHIFT_URL_ANNOTATION, self.module.netlocation, ldapurl_annotation
+            )
+
+        ldapuid_annotation = group["metadata"].get("annotations", {}).get(LDAP_OPENSHIFT_UID_ANNOTATION)
+        if not ldapurl_annotation or ldapurl_annotation != self.module.netlocation:
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/module_utils/openshift_groups.py:210]: checking was wrong, updated.
Thanks

File: molecule/default/roles/openshift_adm_groups/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+              value: "{{ ldap_root }}"
+          ports:
+          - containerPort: 1389
+  register: pod_info
+
+- name: Set Pod Internal IP
+  set_fact:
+    podIp: "{{ pod_info.result.status.podIP }}"
+
+# - name: Create LDAP service
```
Comment (jillr) [inline molecule/default/roles/openshift_adm_groups/tasks/main.yml:40]: Do the commented out tests in this file need to be removed, enabled, or something else?

File: molecule/default/roles/openshift_adm_groups/tasks/rfc2307.yml [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    vars:
+      merge_rfc2307: "{{ configs.out_scope.rfc2307 | combine({'tolerateMemberNotFoundErrors': 'true'}) }}"
+
+  - name: Synchronize Groups - Partially user defined mapping (tolerateMemberNotFoundErrors=true)
+    community.okd.openshift_adm_groups_sync:
+      config: "{{ config_out_of_scope_tolerate_not_found }}"
+    check_mode: yes
+    register: result
+
+  - name: Assert group sync failed due to non-existent member
```
Comment (jillr) [inline molecule/default/roles/openshift_adm_groups/tasks/rfc2307.yml:178]:
```suggestion
  - name: Assert group sync did not fail (tolerateMemberNotFoundErrors=true)
```

**Review by jillr — COMMENTED** [review]
Noting here our conversation the other day from outside of github; I agree that collection users will want to use the LDAPSyncConfig method for client config.

Just a few small comments, thanks!

File: molecule/default/roles/openshift_adm_groups/tasks/main.yml (hunk 5 above)
Comment (abikouo) [reply molecule/default/roles/openshift_adm_groups/tasks/main.yml:40]: this need to be removed, was originally used when LDAP was deployed using a k8s service

**Review by jillr — APPROVED** [review]
LGTM, thanks @abikouo!

**Pull request merged**